src/software/jetson_nano/services/motor.h

```diff
@@ -45,15 +47,58 @@ class MotorService : public Service
         const TbotsProto::DirectControlPrimitive_DirectVelocityControl& local_velocity,
         float dribbler_speed_rpm);
 
+    /**
+     * Trinamic API binding, sets spi_cs_driver_to_controller_demux appropriately
+     * and calls readWriteByte. See C++ implementation file for more info
+     *
+     * @param motor Which motor to talk to (in our case, the chip select)
+     * @param data The data to send
+     * @param last_transfer The last transfer of uint8_t data for this transaction.
+     */
+    uint8_t tmc4671ReadWriteByte(uint8_t motor, uint8_t data, uint8_t last_transfer);
+    uint8_t tmc6100ReadWriteByte(uint8_t motor, uint8_t data, uint8_t last_transfer);
+
    private:
+    /**
+     * Trigger an SPI transfer over an open SPI connection
+     *
+     * @param fd The SPI File Descriptor to transfer data over
+     * @param tx The tx buffer, data to send out
+     * @param rx The rx buffer, will be updated with data from the full-duplex transfer
+     * @param len The length of the tx and rx buffer
+     *
+     */
+    void spiTransfer(int fd, uint8_t const* tx, uint8_t const* rx, unsigned len);
+
+    /**
+     * Trinamic API Binding function
+     *
+     * @param motor Which motor to talk to (in our case, the chip select)
+     * @param data The data to send
+     * @param last_transfer The last transfer of uint8_t data for this transaction.
+     */
+    uint8_t readWriteByte(uint8_t motor, uint8_t data, uint8_t last_transfer);
+
+    // Select between driver and controller gpio
+    GPIO spi_cs_driver_to_controller_demux_gpio;
+
+    // Enable driver gpio
+    GPIO driver_control_enable_gpio;
+
+    // Transfer Buffers
+    uint8_t tx[5] = {0};
+    uint8_t rx[5] = {0};
+
+    // Transfer State
```

Comment:
Are these supposed to be like SPI semaphores?

Reply:
those are mostly just ways of creating an implicit fsm because of how trinamic sends back data.

we set transfer_started and one of currently_writing or reading to true, and then we process data until last_transfer is true, which case we reset everthing.